File: csharp/LoanServer/LoanRepository.cs
using System;
using System.IO;
using System.Text;

namespace MikadoMethod.LoanServer {
  public class LoanRepository {

    public const string REPOSITORY_ROOT = "./loan";
    public const string FILE_EXTENSION = ".loan";

    public static LoanApplication Fetch(string ticketId) {
      long id;
      long.TryParse(ticketId, out id);
      return Fetch(id);
    }

    public static LoanApplication Fetch(long ticketId) {
      try {
        var json = File.ReadAllText(FileFromApplication(ticketId), Encoding.UTF8);
        return Json.From<LoanApplication>(json);
      } catch(FileNotFoundException e) {
        throw new ApplicationException("Ticket not found", e);
      } catch(Exception e) {
        throw new ApplicationException("Could not fetch application", e);
      }
    }

    public static Ticket Store(LoanApplication application) {
      try {
        Directory.CreateDirectory(REPOSITORY_ROOT);
        var json = Json.To<LoanApplication>(application);
        File.WriteAllText(FileFromApplication(application.ApplicationNo), json, Encoding.UTF8);
      } catch(Exception e) {
        throw new ApplicationException("Could not store application", e);
      }
      return new Ticket{ Id = application.ApplicationNo };
    }

    private static string FileFromApplication(long applicationNo) {
      return REPOSITORY_ROOT + "/" + applicationNo + FILE_EXTENSION;
    }

    public static Ticket Approve(string ticketId) {
      var application = Fetch(ticketId);
      application.Approved = true;
      Store(application);
      return new Ticket{ Id = application.ApplicationNo };
    }

  }
}

File: cpp/loan_server/source/LoanRepository.cc
#include <loan_server/LoanRepository.h>

#include <loan_server/LoanApplication.h>
#include <loan_server/Ticket.h>

#include <fstream>

namespace mikado
{
  namespace
  {
    static std::string const FILE_EXTENSION = ".loan";

    std::string file_from_application(long application_no) {
      return LoanRepository::REPOSITORY_ROOT + "/"
          + std::to_string(application_no)
          + FILE_EXTENSION;
    }
  }

  std::string const LoanRepository::REPOSITORY_ROOT = ".";

  LoanApplication LoanRepository::fetch(std::string ticket_id) {
    return fetch(stol(ticket_id));
  }

  LoanApplication LoanRepository::fetch(long ticket_id) {
    std::ifstream file{file_from_application(ticket_id)};
    Json json;
    file >> json;
    LoanApplication application = json;
    return application;
  }

  Ticket LoanRepository::store(LoanApplication const & application) {
    std::ofstream file{file_from_application(application.application_no)};
    file << Json(application);
    return Ticket{application.application_no};
  }

  Ticket LoanRepository::approve(std::string ticket_id) {
    auto application = fetch(ticket_id);
    application.approved = true;
    store(application);
    return Ticket{application.application_no};
  }
}

